vault/plugin_catalog.go

```diff
@@ -59,12 +84,98 @@ func (c *Core) setupPluginCatalog(ctx context.Context) error {
 	return nil
 }
 
+func (c *PluginCatalog) removeMultiplexedClient(ctx context.Context, name, id string) {
+	if _, ok := c.multiplexedClients[name]; !ok {
+		return
+	}
+
+	delete(c.multiplexedClients[name].connections, id)
+	c.logger.Debug("deleted multiplexedClients connection entry", "id", id)
+
+	if len(c.multiplexedClients[name].connections) == 0 {
+		// TODO(JM): This leaves child process behind after vault exits
+		c.multiplexedClients[name].client.Kill()
+		delete(c.multiplexedClients, name)
+		c.logger.Debug("deleted multiplexedClients plugin entry", "name", name)
+	}
+}
+
+func (c *PluginCatalog) getMultiplexedClient(pluginName string) *MultiplexedClient {
+	if mpc, ok := c.multiplexedClients[pluginName]; ok {
+		c.logger.Debug("MultiplexedClient exists", "pluginName", pluginName)
+
+		return mpc
+	}
+
+	c.logger.Debug("MultiplexedClient does not exist", "pluginName", pluginName)
+
+	return c.newMultiplexedClient(pluginName)
+}
+
+func (c *PluginCatalog) newMultiplexedClient(pluginName string) *MultiplexedClient {
+	if c.multiplexedClients == nil {
+		c.multiplexedClients = make(map[string]*MultiplexedClient)
+		c.logger.Debug("created multiplexedClients map")
+	}
+
+	mpc := &MultiplexedClient{
+		connections: make(map[string]plugin.ClientProtocol),
+	}
+
+	// set the MultiplexedClient for the given plugin name
+	c.multiplexedClients[pluginName] = mpc
+	c.logger.Debug("set the MultiplexedClient for", "pluginName", pluginName)
+
+	return mpc
+}
+
+// getPluginClient returns a client for managing the lifecycle of a plugin
+// process and an ID for a newly created entry in the MultiplexedClients'
+// connections map.
+func (c *PluginCatalog) getPluginClient(ctx context.Context, sys pluginutil.RunnerUtil, pluginRunner *pluginutil.PluginRunner, namedLogger log.Logger, isMetadataMode bool) (plugin.ClientProtocol, string, error) {
```

Comment:
I'm not sure we need to pass sys or pluginRunner into this function, our Core object should be able to reference or derive these itself. Maybe it just needs the plugin name we want to create?

Reply:
Thanks! I will take a closer look

Reply:
I have looked into this for a while today. I am not sure we can omit pluginRunner since this function is also in the path of registering plugins and we haven't yet written the plugin entry to storage.